lib/host/wasi/inode-win.cpp

```diff
@@ -82,53 +533,308 @@ INode INode::stdErr() noexcept {
+  // open the file in the win32 API hence it will be better to retrieve the
+  // saved flags used during the file open
+  // FIXME: Find a better way
+  FdStat.fs_flags = (*SavedFdFlags);
+
+  return {};
 }
 
-WasiExpect<void> INode::fdFdstatSetFlags(__wasi_fdflags_t) const noexcept {
-  return WasiUnexpect(__WASI_ERRNO_NOSYS);
+WasiExpect<void>
+INode::fdFdstatSetFlags(__wasi_fdflags_t FdFlags) const noexcept {
+  // The __WASI_FDFLAGS_APPEND flag is ignored as it cannot be changed for an
+  // open file
+
+  DWORD Attributes = FILE_ATTRIBUTE_NORMAL;
+
+  FILE_BASIC_INFO BasicInfo;
+
+  // Source: https://devblogs.microsoft.com/oldnewthing/20210729-00/?p=105494
+  if ((FdFlags & __WASI_FDFLAGS_SYNC) || (FdFlags & __WASI_FDFLAGS_RSYNC)) {
+    // Linux does not implement RSYNC and glibc defines O_RSYNC as O_SYNC
+    Attributes |= FILE_FLAG_WRITE_THROUGH | FILE_FLAG_NO_BUFFERING;
+  }
+  if (FdFlags & __WASI_FDFLAGS_DSYNC) {
+    Attributes |= FILE_FLAG_NO_BUFFERING;
+  }
+  if (FdFlags & __WASI_FDFLAGS_NONBLOCK) {
+    Attributes |= FILE_FLAG_OVERLAPPED;
+  }
+
+  if (unlikely(GetFileInformationByHandleEx(Handle, FileBasicInfo, &BasicInfo,
+                                            sizeof(BasicInfo))) == FALSE) {
+    return WasiUnexpect(fromWinError(GetLastError()));
+  }
+
+  // Update the attributes
+  BasicInfo.FileAttributes = Attributes;
+
+  if (unlikely(SetFileInformationByHandle(Handle, FileBasicInfo, &BasicInfo,
+                                          sizeof(BasicInfo))) == FALSE) {
+    return WasiUnexpect(fromWinError(GetLastError()));
+  }
+
+  return {};
 }
 
-WasiExpect<void> INode::fdFilestatGet(__wasi_filestat_t &) const noexcept {
-  return WasiUnexpect(__WASI_ERRNO_NOSYS);
+WasiExpect<void>
+INode::fdFilestatGet(__wasi_filestat_t &FileStat) const noexcept {
+  // TODO: Complete this partially implemented function after finding equivalent
+  // flags/attributes for __wasi_filetype_t.
+
+  // Update the File information
+  FileInfo.emplace();
+  if (unlikely(GetFileInformationByHandle(Handle, &*FileInfo) == FALSE)) {
+    return WasiUnexpect(fromWinError(GetLastError()));
+  }
+
+  FileStat.filetype =
+      fromFileType((*FileInfo).dwFileAttributes, GetFileType(Handle));
+
+  // Windows does not have an equivalent for the INode number.
+  // A possible equivalent could be the File Index
+  // Source:
+  // https://stackoverflow.com/questions/28252850/open-windows-file-using-unique-id/28253123#28253123
+  // this
+  // https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-openfilebyid?redirectedfrom=MSDN
+  // and this
+  // https://docs.microsoft.com/en-us/windows/win32/api/winbase/ns-winbase-file_id_info
+  FileStat.ino = 0;
+  // TODO: Find an equivalent for device ID in windows
+  FileStat.dev = 0;
+  FileStat.nlink = (*FileInfo).nNumberOfLinks;
+  FileStat.size = static_cast<uint64_t>((*FileInfo).nFileSizeLow) +
+                  (static_cast<uint64_t>((*FileInfo).nFileSizeHigh) << 32);
```

Comment:
What's the type of `nFileSizeHigh`? Should we cast it first, then do the shift?

Reply:
`nFileSizeHigh` is a `DWORD` a 32 bit signed integer. Yes, we have to cast it first. By casting it first we are converting it into a 64 bit integer which occupies the less significant half of the integer. So we have to shift it to the left by 32 bits.